torch/csrc/jit/test_jit.cpp

```diff
@@ -536,6 +539,74 @@ void testADFormulas() {
   }
 }
 
+autograd::Variable var(at::Type & t, at::IntList sizes, bool requires_grad) {
+  return autograd::make_variable(t.rand(sizes), requires_grad);
+}
+autograd::Variable undef() {
+  return autograd::Variable();
+}
+
+int device(const autograd::Variable & v) {
+  return v.type().is_cuda() ? v.get_device() : -1;
+}
+
+bool isEqual(at::IntList lhs, at::IntList rhs) {
+  return lhs.size() == rhs.size() && std::equal(lhs.begin(), lhs.end(), rhs.begin());
+}
+
+bool isEqual(const TensorInfo & ti, const autograd::Variable & v) {
+  if(!ti.defined())
+    return ti.defined() == v.defined();
+  return
+    ti.device() == device(v) &&
+    ti.requires_grad() == v.requires_grad() &&
+    ti.type() == v.type().scalarType() &&
+    isEqual(ti.sizes(), v.sizes()) &&
+    isEqual(ti.strides(), v.strides());
+}
+
+void argumentSpecTest() {
+  auto & CF = at::CPU(at::kFloat);
+  auto & CD = at::CPU(at::kDouble);
+  auto & GF = at::CUDA(at::kFloat);
+  auto & GD = at::CUDA(at::kDouble);
+
+  autograd::variable_list list =  { var(CF, {1}, true), var(CD, {1, 2}, false) , var(GF, {}, true), var(GD, {4,5,6}, false), undef()};
+  // make sure we have some non-standard strides
+  list[1].transpose_(0, 1);
+
+  // same list but different backing values
+  autograd::variable_list list2 = { var(CF, {1}, true), var(CD, {1, 2}, false) , var(GF, {}, true), var(GD, {4,5,6}, false), undef()};
+  list2[1].transpose_(0, 1);
+
+
+  ArgumentSpec a(false, list);
+  ArgumentSpec b(false, list);
+  JIT_ASSERT(a.hashCode() == b.hashCode());
+
+  JIT_ASSERT(a == b);
+  JIT_ASSERT(ArgumentSpec(false, list2) == a);
+  for(size_t i = 0; i < list.size(); ++i) {
+    JIT_ASSERT(isEqual(a.tensorInfo(i), list[i]));
+  }
+  ArgumentSpec no_grad(/*nograd=*/true, list);
+  JIT_ASSERT(!(no_grad == a));
```

Comment:
can we please add `operator!=` implemented as a simple negation of equality?

Reply:
C++ generates so many default methods... Why can't it generate this one :)

Reply:
you don't use what you don't pay for, right? :wink: